smll/FaceDetector.cpp

```diff
@@ -146,8 +146,50 @@ namespace smll {
 			(detect.getStride() * cropInfo.offsetY) +
 			(detect.getNumElems() * cropInfo.offsetX);
 
-		cv::Mat gray(cropInfo.height, cropInfo.width, CV_8UC1, cropData, m_detect.getStride());
-		currentImage = gray;
+		switch (detect.type) {
+		case IMAGETYPE_GRAY:
+		{
+			cv::Mat gray(cropInfo.height, cropInfo.width, CV_8UC1, cropData, m_detect.getStride());
+			currentImage = gray;
```

Comment:
`currentImage = gray` is an assignment operation, it doesn't copy any data from `gray` to `currentImage` but keeps a reference to the current `gray` Mat. This can become problem if we tweak gray in the middle or the memory specific to gray gets corrupted. `.clone()` on the other side is a deep copy of all the contents from `gray` Mat to `currentImage` mat. It is a good practice to use `.clone()` than assignment operators for real-time and possibly concurrent applications.

Reply:
Please note that the assignment operator for OpenCV Mat is quite different from std::vector's assignment operator. 
Links:
[1] std::vector - http://www.cplusplus.com/reference/vector/vector/operator=/
[2] cv::Mat - https://docs.opencv.org/3.1.0/d3/d63/classcv_1_1Mat.html#aed1f81fe7efaacc2bd95149cdfa34302
[3] https://stackoverflow.com/questions/6411476/opencv-matoperator-does-it-support-copy-on-write

Reply:
Ok. Where do we modify gray in this case?

Reply:
We are not modifying the `gray` Mat here. And gray is not being used anywhere outside the `computeCurrentImage` method. What is the system's garbage collector deletes gray even before we start doing face detection based Ops on currentImage

Reply:
That is why I mentioned Mat's reference counter. Mat has a reference counter that is increased with each assignment and the object is only destroyed once reference counter is 0, once all the assignments are no longer being used. It works so far, we can test it more before release, but this should work fine.

Reply:
Correct me if I am wrong, but reference counts usually harm the performance of multi-threaded applications. You can try doing in single-shot instead of writing the `.clone()`:
```
cv::cvtColor(rgbImage, currentImage, cv::COLOR_RGB2GRAY);
```
instead of
```
cv::Mat gray; cv::cvtColor(rgbImage, gray, cv::COLOR_RGB2GRAY);
currentImage = gray.clone();
```

Reply:
Deep copy trumps an already implemented reference counter. If there are clashes, we can always add clone later.